src/mds/Locker.cc

```diff
@@ -4035,13 +4035,6 @@ void Locker::_do_cap_release(client_t client, inodeno_t ino, uint64_t cap_id,
                   new C_Locker_RetryCapRelease(this, client, ino, cap_id, mseq, seq));
     return;
   }
-  if (seq != cap->get_last_issue()) {
-    dout(7) << " issue_seq " << seq << " != " << cap->get_last_issue() << dendl;
-    // clean out any old revoke history
-    cap->clean_revoke_from(seq);
-    eval_cap_gather(in);
-    return;
-  }
```

Comment:
Can you explain why dropping this block and doing a full cap release is better? The commit outlines the problem, but not how this is the solution.

Also, this code is old and originates from fixing a snapshot caps bug (3962a7510fd7f963c98ee34118cdd15cc9014c10), so I'm a bit nervous about changes to it.

Reply:
The commit (3962a7510fd7f963c98ee34118cdd15cc9014c10) was added by Zheng and just copied from `Locker::handle_client_cap_release()` to `Locker::_do_cap_release()` to defer releasing the caps.

From Sage's old commit (b180b6ce1981b6d23b68273266a0ab6594f40824) the release and caps revocation could be racy, the `seq != cap->get_last_issue()` could be true. And then it will call  `cap->clean_revoke_from(seq);`.

But in `cap->clean_revoke_from(seq);`:

```
215   void clean_revoke_from(ceph_seq_t li) {
216     bool changed = false;      
217     while (!_revokes.empty() && _revokes.front().last_issue <= li) {                                                                               
218       _revokes.pop_front();    
219       changed = true;
220     }
221     if (changed) {
222       bool was_revoking = (_issued & ~_pending);
223       calc_issued();           
224       if (was_revoking && _issued == _pending) {
225         item_revoking_caps.remove_myself(); 
226         item_client_revoking_caps.remove_myself();
227         maybe_clear_notable(); 
228       }
229     }
230   }
```
The `_revokes.front().last_issue <= li` could be false and then the revoking item won't be removed.

And then back in `Locker::_do_cap_release()` it will just return and skip calling the following `remove_client_cap(in, cap);`, which will clean the revoking item too.

And then the revoking item is kept.

Once the `MDS` get a cap release request from clients, it means the cap in client side has been removed already. Then will the `MDS` could get another cap release request for the same cap ? If not then what is the `MDS` waiting for ?

And also from another user's feedback once the they removed the corresponding directory this WARN disappeared, just closing it won't work.

Reply:
What about this scenario:

```
               mds                             client
                                               - Releases cap and put Inode
      - Issues new cap update (new issued)
      - Increase cap->seq and sends
        revokes req to the client
      - receives release, removes client cap
        (your change)
                                               - Adds cap
                                               - Does stuff (DANGER!) 
                                               - Receives revoke & issues cap release
      - drops bogus release?                   
```

I may be forgetting a detail here which would avoid this problem.

Reply:
From reading the client side code, once the caps is added to the `release` list the `inode` will be released just after that. 

So when the clients received the new cap requests after the `inode` being released the clients will just trigger to flush the `release` list and then drop the requests.

So it's safe.

Reply:
Here right?

https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L5254-L5258

as long as the kclient has the same behavior then I think this changeset is good.

Reply:
yeah, correct.